### DISCUSSION TITLE
Where to focus if I want to extend Magma to support more management protocols

### DISCUSSION THREAD

**Discussion by feng-zhe** [magma/magma#6083] [Q&A] [answered]
Hi experts,

I am new to Magma and I wondering how to extend Magma to support protocols other than TR-069 (CWMP). I assume I should focus on the code in `magma/lte/gateway/` but I am wondering if there is any doc for this kind of development.

Thanks.

**Comment by hcgatewood** [comment]
cc @ardzoht @andreilee

**Comment by andreilee** [comment]
Hi Feng Zhe,

We don't have a doc for this kind of development. Are there management protocols in particular that you are interested in? It can be difficult to give a good answer without more specifics.

**Comment by karthiksubraveti** [comment]
Adding @ulaskozat as well.

**Comment by feng-zhe** [comment]
Hi @andreilee, thanks for you reply. We are currently using SNMP to manage our eNB. And its firmware also supports a set of rest APIs for management. Not sure if this is a common case but that is how we are managing them now.

Currently we are exploring if and how AGW can support eNBs of this kind. Thanks.

**Comment by andreilee** [comment, marked as answer]
Outside of TR-069, Magma can also work with eNBs that it does not directly manage the configuration of.
cc @ardzoht for more details.

If you're sure you want to extend Magma for more management protocols, the most natural place to add this code would be the [enodebd](https://github.com/magma/magma/tree/master/lte/gateway/python/magma/enodebd) service, which currently handles all TR-069 messaging.

The REST API for eNodeB configuration is defined in [`swagger.v1.yml`](https://github.com/magma/magma/blob/master/lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml#L2184-L2200). It's the responsibility of `enodebd` service to take that desired configuration and uses TR-069 to get the specified eNB into that state. `enodebd` is thus a natural place to add code to support more management protocols.

I'd also be happy to chat more over a call.

**Reply by feng-zhe** [reply]
Hi andrielee, thank you for you information! Will take a look and see what I can do. Thanks agani!

**Comment by ardzoht** [comment]
Hi @feng-zhe, as @andreilee mentioned, we do have support for configuring eNodeb devices that are not configured on the gateway itself, for this type of configuration, you can take a look at https://magma.github.io/magma/docs/lte/enodebd#manual-configuration where it specifies how to register an externally-managed eNB and the steps to setup the gateway to enable this support.

**Reply by feng-zhe** [reply]
Thanks, ardzoht! The pain points we have is that manually config the eNBs (via SNMP or HTTPS) is hard to do. Will try to connect manually configured eNB first and see if AGW can be extended to auto-config them. Thanks!
